pip/index.py

```diff
@@ -376,16 +381,35 @@ def mkurl_pypi_url(url):
                     version.location.url for version in dependency_versions
                 ])
             )
+
         file_versions = list(
             self._package_versions(
                 [Link(url) for url in file_locations],
                 req.name.lower()
             )
         )
-        if (not found_versions and not
-                page_versions and not
-                dependency_versions and not
-                file_versions):
+        if file_versions:
+            file_versions.sort(reverse=True)
+            logger.debug(
+                'Local files found: %s',
+                ', '.join([
+                    url_to_path(candidate.location.url)
+                    for candidate in file_versions
+                ])
+            )
+
+        # This is an intentional priority ordering
```

Comment:
there's 3 tests related to order that could/should be rewritten to use this refactor.

`test_finder_priority_file_over_page`
`test_finder_priority_page_over_deplink`
`test_finder_priority_nonegg_over_eggfragments`